libs/x-content/src/main/java/org/elasticsearch/xcontent/support/filtering/FilterPath.java

```diff
@@ -11,102 +11,194 @@
+     * @param name the xcontent property name
+     * @param nextFilters nextFilters is a List, used to check the inner property of name
+     * @return true if the name equal a final node, otherwise return false
+     */
+    boolean matches(String name, List<FilterPath> nextFilters) {
+        if (nextFilters == null) {
+            return false;
         }
-        return null;
-    }
 
-    public boolean matches() {
-        return next == null;
-    }
+        FilterPath termNode = termsChildren.get(name);
+        if (termNode != null) {
+            if (termNode.isFinalNode()) {
+                return true;
+            } else {
+                nextFilters.add(termNode);
+            }
+        }
 
-    public boolean isDoubleWildcard() {
-        return doubleWildcard;
-    }
+        for (FilterPath wildcardNode : wildcardChildren) {
+            String wildcardPattern = wildcardNode.getPattern();
+            if (Glob.globMatch(wildcardPattern, name)) {
+                if (wildcardNode.isFinalNode()) {
+                    return true;
+                } else {
+                    nextFilters.add(wildcardNode);
+                }
+            }
+        }
 
-    public boolean hasDoubleWildcard() {
-        if (filter == null) {
-            return false;
+        if (isDoubleWildcard) {
+            nextFilters.add(this);
         }
-        return filter.indexOf("**") >= 0;
-    }
 
-    public boolean isSimpleWildcard() {
-        return simpleWildcard;
+        return false;
     }
 
-    public String getSegment() {
-        return segment;
-    }
+    private static class FilterPathBuilder {
+        private class BuildNode {
+            private final Map<String, BuildNode> children;
+            private final boolean isFinalNode;
 
-    public FilterPath getNext() {
-        return next;
-    }
+            BuildNode(boolean isFinalNode) {
+                children = new HashMap<>();
+                this.isFinalNode = isFinalNode;
+            }
+        }
 
-    public static FilterPath[] compile(Set<String> filters) {
-        if (filters == null || filters.isEmpty()) {
-            return null;
+        private BuildNode root = new BuildNode(false);
+
+        void insert(String filter) {
+            insertNode(filter, root);
         }
 
-        List<FilterPath> paths = new ArrayList<>();
-        for (String filter : filters) {
-            if (filter != null) {
-                filter = filter.trim();
-                if (filter.length() > 0) {
-                    paths.add(parse(filter, filter));
+        FilterPath build() {
+            return buildPath("", root);
+        }
+
+        void insertNode(String filter, BuildNode node) {
+            int end = filter.length();
+            for (int i = 0; i < end;) {
+                char c = filter.charAt(i);
+                if (c == '.') {
+                    String field = filter.substring(0, i).replaceAll("\\\\.", ".");
```

Comment:
Losing the `hasEscape` flag that your non-builder version had seems to really slow down the builder version.